Blog post review (#1)
datawire/telepresence-compose-demo: BLOG.md

# Telepresence with Docker Compose

> When it comes to developing an application, one of the potential friction points is the developer's machine. Several issues may arise:

Reviewer: ```suggestion
When it comes to developing an application, one of the potential friction points is your local setup. Several issues may arise:
```

- Tight security policies may be in place.
- The machine's operating system may differ significantly from the server's.
- Each individual has their own setup and way of working.

To address this challenge, platform engineers or developers themselves often resort to using a common solution that allows anyone to quickly start working on a project.

> One commonly seen tool for this purpose is Docker, and more specifically, Docker Compose.

Reviewer: ```suggestion
One common solution for this problem, is Docker, and more specifically, Docker Compose.
```

> Both Compose and Telepresence assist developers in working on their applications, although they serve different purposes:

Reviewer: ```suggestion
Both Docker Compose and Telepresence assist developers in working on their applications, although they serve different purposes:
```

> - Compose provides a highly customizable way of running an entire application locally.

Reviewer: ```suggestion
- Docker Compose provides a highly customizable way of running an entire application stack locally.
```

> - Telepresence enables running a local environment within a remote cluster.

Reviewer: ```suggestion
- Telepresence enables combining your local environment with a remote cluster.
```

> Let's explore how to harmonize and combine these two approaches.

Reviewer: ```suggestion
Let's explore how to harmonize these two approaches.
```

## A simple example

> Let's start with a simple example of infrastructure, and consider the following application :

Reviewer: ```suggestion
Let's start with a simple example and consider the following application :
```

- A contact-api micro service,
- A Postgres database.

The user API features an endpoint that retrieves user data, and in order to obtain the country information for each user, it makes additional requests to the contact API.

These microservices store their data in a PostgreSQL database.

Here is a schema summarizing the infrastructure:

![Screenshot](images/compose-schema.png)

And the equivalent docker-compose file.

```yaml
services:
  contactapi:
    build:
      context: .
      dockerfile: docker/contactapi/Dockerfile
    environment:
      DATASET: local
      DB_HOST: postgres
      DB_USERNAME: postgres
      DB_PASSWORD: postgresql
      PORT: 8081
    depends_on:
      - userapi
    ports:
      - 8081:8081

  userapi:
    build:
      context: .
      dockerfile: docker/userapi/Dockerfile
    environment:
      DATASET: local
      DB_HOST: postgres
      DB_USERNAME: postgres
      DB_PASSWORD: postgresql
      CONTACT_API_URL: "http://contactapi:8081"
      PORT: 8080
    depends_on:
      - postgres
    ports:
      - 8080:8080

  postgres:
    image: postgres:15-bullseye
    environment:
      POSTGRES_PASSWORD: postgresql
    ports:
      - 5432:5432
```

> You can run it locally with compose by running:

Reviewer: ```suggestion
You can start your services locally with compose by running:
```

```cli
git clone https://github.com/datawire/telepresence-compose-demo
cd telepresence-compose-demo
docker compose up
```

> And you should be able to see your service locally :

Reviewer: ```suggestion
You can test your service with:
```

```cli
curl localhost:8080/users
```

As you can see, it only contains one test user, that's the local database.

Press `Ctrl + C` to shutdown the Compose run.

## Going Kubernetes

> Now let's say that you have an equivalent Kubernetes infrastructure.

Reviewer: ```suggestion
Now, lets assume you have the equivalent services running within Kubernetes.
```

> You can find its definition in the repository used above, and create it with:

Reviewer: ```suggestion
You can find the configuration at: <insert link to files here>, and create it with:
```

Reviewer: Since this is a blogpost we want to be as direct as possible so we should include links instead of saying "found somewhere in what we listed somewhere else"

```cli
# Create a namespace (feel free to change its name)
kubectl create namespace telepresence-compose-demo
kubectl apply -f infra -n telepresence-compose-demo
```

> And this is what we get:

Reviewer: ```suggestion
And this is what the resulting infrastructure:
```

userapi      ClusterIP   10.43.154.149   <none>        80/TCP     6h16m

> kubectl get pods -n telepresence-compose-demo
NAME                          READY   STATUS    RESTARTS   AGE
postgres-7976596b89-wfnp6     1/1     Running   0          6h16m
contactapi-7d8688b987-4hk2l   1/1     Running   0          6h16m
userapi-687ffc88b6-bfthc      2/2     Running   0          4h59m

> telepresence connect -n telepresence-compose-demo --docker
> telepresence list
contactapi: ready to intercept (traffic-agent not yet installed)
postgres  : ready to intercept (traffic-agent not yet installed)
userapi   : ready to intercept (traffic-agent not yet installed)
```

How can you test your compose file against these services ?

Typically, you have two options: either you continuously test everything locally, ensuring it closely resembles the real setup, or you maintain a separate test environment in a different namespace where you can update the pod with your test version.

This is where Telepresence can be extremely beneficial for your setup, particularly because it offers integration with Docker Compose.

## The Telepresence specification

### Basic knowledge

> Telepresence supports a file format that it is possible to use to persist any type of configuration: The Intercept Specification.

Reviewer: ```suggestion
Telepresence supports a YAML document that allows you to create a consistent  and repeatable environment: The Intercept Specification.
```

It has three sections :

> - `connection`: This one is optional, but it allows you to define which context the spec should be used with.

Reviewer: ```suggestion
- `connection`: This one is optional, but it allows you to define which Kubernetes context the spec should be used with.
```

- `workloads`: It defines the workloads to intercept, and how (header, ports, ...).

> - `handlers`: That's who you'll handle the traffic from these intercepts.

Reviewer: ```suggestion
- `handlers`: The local service that will handle the intercepted traffic from the workload.
```

Consider this specification file:

```yaml
connection:
  namespace: telepresence-compose-demo # Consider your local setup moved to this namespace.
workloads:
  - name: userapi
    intercepts:
      - handler: userapi # Matches the handler's name.
        global: true
        previewURL: false
        port: 80 # The corresponding Kubernetes service uses the port 80.
        localPort: 8080 # The local container uses port 8080.
handlers:
  - name: userapi
    docker:
      build:
        context: ./
        args: # Build the docker image on the fly.
          - --file
          - ./docker/userapi/Dockerfile
```

Setting the namespace in the connection field, means that any container you'll define in this file, will be
able to communicate with the cluster as if it was actually located there.

We indicate that the `userapi` workload needs to be intercepted. It means that any request going to this service
will be redirected to your local container (we call that a global intercept, but you could also do it for a subset of the

> traffic using a personal intercepts).

Reviewer: Maybe add a link here to our intercepts docs

Then we define the handler that will receive the traffic. In this case, we use `userapi`'s Dockerfile.

Now let's run that spec:

```cli
> telepresence intercept run telepresence.spec.yaml
   Intercept name    : userapi
   State             : ACTIVE
   Workload kind     : Deployment
   Destination       : 127.0.0.1:8080
   Volume Mount Point: /var/folders/x_/4x_4pfvx2j3_94f36x551g140000gp/T/telfs-3467079569
   Intercepting      : all TCP requests
   Layer 5 Hostname  : userapi.telepresence-compose-demo.svc.cluster.local
Intercept spec "telepresence.spec" started successfully, use ctrl-c to cancel.
Attempt 1: Checking PostgreSQL connection...
PostgreSQL is ready!
2023/11/01 20:11:24 Listening on 8080
```

> If you hit the real endpoint, you will see the traffic in the logs. You can get access it

Reviewer: Which logs, the pod logs or local logs? 🙂 we should be as specific as we can so users know exactly where to look. It can feel silly writing that way since it seems obvious but its integral in posts like these where someone might know absolutely nothing about Kubernetes, or they may be an expert, either way we need to be super specific

through its external IP:

```cli
> kubectl get svc userapi -n telepresence-compose-demo
NAME      TYPE        CLUSTER-IP    EXTERNAL-IP    PORT(S)   AGE
userapi   ClusterIP   10.43.25.37   x.x.x.x        80/TCP    56m
```

> If you don't have any load balancer to expose your service, you can still try to

Reviewer: ```suggestion
If you don't have any load balancer to expose your service, you can still ```
